<h1>Data Cleaning / Munging</h1>

<p>In these notes we look at creating various plots based on a set of data.  We have a datafile called degreeClassification.csv (available on blackboard) which consists of the following columns:</p>
<ul>
<li>Academic Year:  The year that the degree was awarded.</li>
<li>Mode of study: Full-time, part-time or both.</li>
<li>Sex: Male, Female, Other or all.</li>
<li>Country of HE provider: England, Northern Ireland, Scotland, Wales or all.</li>
<li>Classification of First Degree: First, Upper Second, Lower Second or Third/Pass.</li>
<li>Number: The number of students achieving that degree.</li>
<li>Percentage:  The percentage of students achieving that degree.</li>
</ul>   

<p>We start the exercises by getting the data ready to be plotted</p>

<h2>Import PANDAs</h2>

In [1]:
import pandas as pd

<h2>Read the data into a PANDAs DataFrame</h2>
<p>Read the 'degreeClassification.csv' in to a data frame called df, use the head method to output the first 20 rows of the datafile so that the structure of the dataframe can be explored.</p>

In [2]:
df = pd.read_csv("degreeClassification.csv")
df.head()

Title  \
0                  Location   
1            Academic years   
2               Data source   
3          Data source link   
4  Data file canonical link   

  Percentage of first degree qualifiers obtaining each classification by sex, mode of study and country of HE provider  \
0                                                 UK                                                                     
1                                 2013/14 to 2017/18                                                                     
2                                               HESA                                                                     
3  https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
4  https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     

  Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  
0        NaN        NaN        NaN        NaN        NaN  
1        NaN        NaN        NaN        NaN        NaN  
2        NaN        NaN        NaN        NaN        NaN  
3        NaN        NaN        NaN        NaN        NaN  
4        NaN        NaN        NaN        NaN        NaN

<h2>Skipping Summary Rows</h2>
<p>Read the datafile and skip the rows that contain summary information.  Use the head method and output the first 20 rows.  The column names should be:
<ul>    
<li>Academic Year</li>
<li>Mode of study</li>
<li>Sex</li>
<li>Country of HE provider</li>
<li>Classification of First Degree</li>
<li>Number</li>
<li>Percentage</li>
</ul>    
</p>

In [11]:
df = pd.read_csv("degreeClassification.csv", skiprows = 17)
df.head()

Academic Year Mode of study  Sex Country of HE provider  \
0       2013/14           All  All                    All   
1       2013/14           All  All                    All   
2       2013/14           All  All                    All   
3       2013/14           All  All                    All   
4       2013/14           All  All                    All   

  Classification of First Degree  Number Percentage  
0                          First   79410        20%  
1                   Lower second   95965        24%  
2                     Third/Pass   21905         6%  
3               Total classified  395605        NaN  
4                   Upper second  198320        50%

<h2>Unique values in a Column</h2>
<p>Output the unique values for the column 'Mode of study' to give a general idea of the categorical data for this column</p>

In [16]:
arrOfUniqueValuesInAColumn = df["Mode of study"].unique()
for uniqueValue in arrOfUniqueValuesInAColumn:
    print(uniqueValue)

All
Full-time
Part-time


<h2>Boolean Masks</h2>

<p>Create a boolean mask and then filter the data based on the categorical data value 'All' for 'Mode of Study'.  We do this in stages to help understand how the more complex aggregated instruction is built.</p>
<p>The overall goal is to apply a mask so that all the rows that contain values part-time or full-time are dropped from the dataframe.  We do this so that we don’t double count, i.e. A count in 'All' will also also be a count in either 'Part-time' or 'Full-time'.</p>

<ol>
<li>Create a mask to output True/False for each row, True if 'Mode of study' is equal to 'All' and False otherwise</li>

In [17]:
df["Mode of study"] == "All"

0        True
1        True
2        True
3        True
4        True
        ...  
1375    False
1376    False
1377    False
1378    False
1379    False
Name: Mode of study, Length: 1380, dtype: bool

<ol start='2'>
<li>Assign the Boolean mask to a variable named mask</li>

In [19]:
mask = df["Mode of study"] == "All"
mask

0        True
1        True
2        True
3        True
4        True
        ...  
1375    False
1376    False
1377    False
1378    False
1379    False
Name: Mode of study, Length: 1380, dtype: bool

<ol start='3'>
<li>Use the variable named mask to update the dataframe named df so that the dataframe is filtered based on the boolean mask</li>
<li>Output the unique values for the column 'Mode of study' to ensure that 'Full-time' and 'Part-time' have been filtered out (don't exist in the dataframe)</li>

In [27]:
df = df.loc[mask]
df["Mode of study"].unique()

array(['All'], dtype=object)

<ol start='5'>
<li>Output the dataframe</li>

In [26]:
df

Academic Year Mode of study    Sex Country of HE provider  \
0          2013/14           All    All                    All   
1          2013/14           All    All                    All   
2          2013/14           All    All                    All   
3          2013/14           All    All                    All   
4          2013/14           All    All                    All   
...            ...           ...    ...                    ...   
1195       2017/18           All  Other                  Wales   
1196       2017/18           All  Other                  Wales   
1197       2017/18           All  Other                  Wales   
1198       2017/18           All  Other                  Wales   
1199       2017/18           All  Other                  Wales   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
3                  Total classified  395605        NaN  
4                      Upper second  198320        50%  
...                             ...     ...        ...  
1195                          First      10         ..  
1196                   Lower second       0         ..  
1197                     Third/Pass       0         ..  
1198               Total classified      20        NaN  
1199                   Upper second      10         ..  

[475 rows x 7 columns]

<h2>Boolean Masks Continued</h2>

<p>Create a boolean mask and then filter the data based on the categorical data 'All' for 'Sex'</p>
<p>Then create a boolean mask and then filter the data based on the categorical data 'All' for 'Country of HE provider'</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 1380 rows to 25 rows.</p>

In [34]:
maskSex = df["Sex"] == "All"
maskCountry = df["Country of HE provider"] == "All"
df = df.loc[maskSex].loc[maskCountry]
df

Academic Year Mode of study  Sex Country of HE provider  \
0          2013/14           All  All                    All   
1          2013/14           All  All                    All   
2          2013/14           All  All                    All   
3          2013/14           All  All                    All   
4          2013/14           All  All                    All   
275        2014/15           All  All                    All   
276        2014/15           All  All                    All   
277        2014/15           All  All                    All   
278        2014/15           All  All                    All   
279        2014/15           All  All                    All   
550        2015/16           All  All                    All   
551        2015/16           All  All                    All   
552        2015/16           All  All                    All   
553        2015/16           All  All                    All   
554        2015/16           All  All                    All   
825        2016/17           All  All                    All   
826        2016/17           All  All                    All   
827        2016/17           All  All                    All   
828        2016/17           All  All                    All   
829        2016/17           All  All                    All   
1105       2017/18           All  All                    All   
1106       2017/18           All  All                    All   
1107       2017/18           All  All                    All   
1108       2017/18           All  All                    All   
1109       2017/18           All  All                    All   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
3                  Total classified  395605        NaN  
4                      Upper second  198320        50%  
275                           First   81640        22%  
276                    Lower second   85250        23%  
277                      Third/Pass   20340         5%  
278                Total classified  370905        NaN  
279                    Upper second  183680        50%  
550                           First   88890        24%  
551                    Lower second   81595        22%  
552                      Third/Pass   19280         5%  
553                Total classified  376330        NaN  
554                    Upper second  186565        50%  
825                           First  100945        26%  
826                    Lower second   79515        20%  
827                      Third/Pass   19025         5%  
828                Total classified  391875        NaN  
829                    Upper second  192395        49%  
1105                          First  110475        28%  
1106                   Lower second   76210        19%  
1107                     Third/Pass   17830         4%  
1108               Total classified  396940        NaN  
1109                   Upper second  192425        48%

<h2>More Boolean Masks</h2>

<p>We are not interested in rows that show the totals of classification for each year, filter these out.</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 25 rows to 20 rows.</p>

In [40]:
maskFilterOutTotalClassification = df["Classification of First Degree"] != "Total classified"
df = df.loc[maskFilterOutTotalClassification]
df

Academic Year Mode of study  Sex Country of HE provider  \
0          2013/14           All  All                    All   
1          2013/14           All  All                    All   
2          2013/14           All  All                    All   
4          2013/14           All  All                    All   
275        2014/15           All  All                    All   
276        2014/15           All  All                    All   
277        2014/15           All  All                    All   
279        2014/15           All  All                    All   
550        2015/16           All  All                    All   
551        2015/16           All  All                    All   
552        2015/16           All  All                    All   
554        2015/16           All  All                    All   
825        2016/17           All  All                    All   
826        2016/17           All  All                    All   
827        2016/17           All  All                    All   
829        2016/17           All  All                    All   
1105       2017/18           All  All                    All   
1106       2017/18           All  All                    All   
1107       2017/18           All  All                    All   
1109       2017/18           All  All                    All   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
4                      Upper second  198320        50%  
275                           First   81640        22%  
276                    Lower second   85250        23%  
277                      Third/Pass   20340         5%  
279                    Upper second  183680        50%  
550                           First   88890        24%  
551                    Lower second   81595        22%  
552                      Third/Pass   19280         5%  
554                    Upper second  186565        50%  
825                           First  100945        26%  
826                    Lower second   79515        20%  
827                      Third/Pass   19025         5%  
829                    Upper second  192395        49%  
1105                          First  110475        28%  
1106                   Lower second   76210        19%  
1107                     Third/Pass   17830         4%  
1109                   Upper second  192425        48%

<h2>Dropping Columns</h2>
<p>Since columns 'Mode of study', 'Sex', and 'Country of HE provider' don\t provide us with any meaninful information (the all contain the data 'All'), drop these columns from the dataframe.</p>
<p>Output the dataframe to check the columns have been dropped successfully.</p>

In [43]:
df.drop(columns=["Mode of study","Sex", "Country of HE provider"])

Academic Year Classification of First Degree  Number Percentage
0          2013/14                          First   79410        20%
1          2013/14                   Lower second   95965        24%
2          2013/14                     Third/Pass   21905         6%
4          2013/14                   Upper second  198320        50%
275        2014/15                          First   81640        22%
276        2014/15                   Lower second   85250        23%
277        2014/15                     Third/Pass   20340         5%
279        2014/15                   Upper second  183680        50%
550        2015/16                          First   88890        24%
551        2015/16                   Lower second   81595        22%
552        2015/16                     Third/Pass   19280         5%
554        2015/16                   Upper second  186565        50%
825        2016/17                          First  100945        26%
826        2016/17                   Lower second   79515        20%
827        2016/17                     Third/Pass   19025         5%
829        2016/17                   Upper second  192395        49%
1105       2017/18                          First  110475        28%
1106       2017/18                   Lower second   76210        19%
1107       2017/18                     Third/Pass   17830         4%
1109       2017/18                   Upper second  192425        48%

<h2>Checking the Data Type of a Column</h2>
<p>Check the type of data that is stored for one of the values in the percentage column.</p>

In [48]:
type(df["Percentage"][0])

str

<h2>Cleaning and converting to an string data type to a number data type</h2>
<p>The percentage column is a string data type because of the % character.  Remove the % character and convert the values for this column to a numeric data type</p>

In [58]:
# Commented out because it has already been ran
# df["Percentage"] = df["Percentage"].str.replace("%", "").astype(int)

<h2>New Dataframes</h2>
<p>Often it is useful to create seperate smaller dataframes, create four new dataframes named df_first, df_upp_sec, df_low_sec and df_third, to seperate the different degree classifications.</p>

Example:
df_first would output:
<pre>
      Academic Year   Classification of First Degree	Number	 Percentage
0	    2013/14	       First	                    79410	  20%
275	  2014/15	       First	                    81640	  22%
550	  2015/16	       First	                    88890	  24%
825	  2016/17	       First	                    100945	 26%
1105	 2017/18	       First	                    110475	 28%
</pre>

In [65]:
df_first = df.loc[df["Classification of First Degree"] == "First"]
df_upp_sec = df.loc[df["Classification of First Degree"] == "Upper second"]
df_low_sec = df.loc[df["Classification of First Degree"] == "Lower second"]
df_third = df.loc[df["Classification of First Degree"] == "Third/Pass"]


<h2>Save the Dataframes to CSV files</h2>
<p>Save the dataframes df, df_first, df_upp_sec, df_low_sec and df_third to csv files with filesnames that match the dataframe name.

In [66]:
df.to_csv("df.csv")
df_first.to_csv("df_first.csv")
df_upp_sec.to_csv("df_upp_sec.csv") 
df_low_sec.to_csv("df_low_sec.csv")
df_third.to_csv("df_third.csv") 